util/shape-options.hh

```diff
@@ -269,6 +269,186 @@ struct shape_options_t
+
+    /* This is what we do:
+     *
+     * 1. We shape text once. Then segment the text at all the safe-to-concat
+     *    points;
+     *
+     * 2. Then we create two buffers, one containing all the even segments and
+     *    one all the odd segments.
+     *
+     * 3. Because all these segments were safe-to-concat at both ends, we
+     *    expect that concatenating them and shaping should NOT change the
+     *    shaping results of each segment.  As such, we expect that after
+     *    shaping the two buffers, we still get cluster boundaries at the
+     *    segment boundaries, and that those all are safe-to-concat points.
+     *    Moreover, that there are NOT any safe-to-concat points within the
+     *    segments.
+     *
+     * 4. Finally, we reconstruct the shaping results of the original text by
+     *    simply interleaving the shaping results of the segments from the two
+     *    buffers, and assert that the total shaping results is the same as
+     *    the one from original buffer in step 1.
+     */
+
+    hb_buffer_t *fragments[2] {hb_buffer_create_similar (buffer),
+			       hb_buffer_create_similar (buffer)};
+    hb_buffer_t *reconstruction = hb_buffer_create_similar (buffer);
+    hb_segment_properties_t props;
+    hb_buffer_get_segment_properties (buffer, &props);
+    hb_buffer_set_segment_properties (fragments[0], &props);
+    hb_buffer_set_segment_properties (fragments[1], &props);
+    hb_buffer_set_segment_properties (reconstruction, &props);
+
+    unsigned num_glyphs;
+    hb_glyph_info_t *info = hb_buffer_get_glyph_infos (buffer, &num_glyphs);
+
+    unsigned num_chars;
+    hb_glyph_info_t *text = hb_buffer_get_glyph_infos (text_buffer, &num_chars);
+
+    bool forward = HB_DIRECTION_IS_FORWARD (hb_buffer_get_direction (buffer));
+
+    if (!forward)
+      hb_buffer_reverse (buffer);
+
+    /*
+     * Split text into segments and collect into to fragment streams.
+     */
+    {
+      unsigned fragment_idx = 0;
+      unsigned start = 0;
+      unsigned text_start = 0;
+      unsigned text_end = 0;
+      for (unsigned end = 1; end < num_glyphs + 1; end++)
+      {
+	if (end < num_glyphs &&
+	    (info[end].cluster == info[end-1].cluster ||
+	     info[end].mask & HB_GLYPH_FLAG_UNSAFE_TO_CONCAT))
+	    continue;
+
+	/* Accumulate segment corresponding to glyphs start..end. */
+	if (end == num_glyphs)
+	  text_end = num_chars;
+	else
+	{
+	  unsigned cluster = info[end].cluster;
+	  while (text_end < num_chars && text[text_end].cluster < cluster)
+	    text_end++;
+	}
+	assert (text_start < text_end);
+
+	if (0)
+	  printf("start %d end %d text start %d end %d\n", start, end, text_start, text_end);
+
+#if 0
+	hb_buffer_flags_t flags = hb_buffer_get_flags (fragment);
+	if (0 < text_start)
+	  flags = (hb_buffer_flags_t) (flags & ~HB_BUFFER_FLAG_BOT);
+	if (text_end < num_chars)
+	  flags = (hb_buffer_flags_t) (flags & ~HB_BUFFER_FLAG_EOT);
+	hb_buffer_set_flags (fragment, flags);
+#endif
```

Comment:
Not sure what HarfBuzz' practice is for deactivated code. In Blink I would suggest to either move to some deactivatable logging macros, or Skia uses some DEBUGCODE macros IIRC. `if 0` or `#if 0` looks a bit brittle.